### ISSUE TITLE
Issue on Authorized Read in VISSv2-Transport

### ISSUE THREAD

**Issue by wonsuk73** [w3c/automotive#399]
There are read, update, subscribe and unsubscribe operations in Current VISSv2-Transport spec. In addition these operations are seperated with authorized and unauthorized as below.

HTTPS:
5.1.2.1 Read
  5.1.2.1.1 Authorized Read
  5.1.2.1.2 Search Read
  5.1.2.1.3 History Read
  5.1.2.1.4 Service Discovery Read
5.1.2.2 Update
  5.1.2.2.1 Authorized Update

Websocket:
5.2.2.1 Read
  5.2.2.1.1 Authorized Read
  5.2.2.1.2 Search Read
  5.2.2.1.3 History Read
  5.2.2.1.4 Service Discovery Read
5.2.2.2 Update
  5.2.2.2.1 Authorized Update
5.2.2.3 Subscribe
  5.2.2.3.1 Authorized Subscribe
  5.2.2.3.2 Curve Logging Subscribe
5.2.2.4 Unsubscribe

Could we allow a client to access vehicle data without authentication? I guess the car manufacturing company might not want to provide their car data without their control. There might be a few level of authentication to control the data access according to security or privacy level.  Am I right?
If it is right, I would suggest to add the phase for explanation of this context, and to change all of unauthenticated operations with authenticated operations. What's your opinions?

**Comment by wonsuk73** [comment]
@UlfBj, @peterMelco and @tguild Could you review and share your view?

**Comment by SebastianSchildt** [comment]
My (not necessarily correct) understanding at least for WS  was, that "Authorized Read" is the special case, where you can put the whole token into each read request, which might be a good pattern if you want a single value ever hour/day or so, while otherwise it has tremendous overhead as a JWT token likely is  several times the size of the query and the response combined.

I assumed normal reads are only possible after a session using a token has been established. (of course depending on the implementation maybe it opts to allow certain reads without any prior authentication).

**Comment by UlfBj** [comment]
>> Could we allow a client to access vehicle data without authentication? 

We do allow that as the access control is optional. On top of that, even if an OEM supports access control, ch. 8.8 "Access control selection" makes it possible to apply access control to only selected nodes of the tree. So some nodes may then not require a token for access. 

If access control is enabled on a signal, then according to the spec the token MUST be provided in each request to it. 
I do not think we should change that and allow some requests to this signal to be without the token.

**Comment by SebastianSchildt** [comment]
It seems, in light of Ulfs answer,  my "understanding" was wrong.

I would still like to make the observation, that `the token MUST be provided in each request to it`  is a bad design decision as it as soon as an app queries more than 1 value in its lifetime  wastes more than 50% of bandwidth on redundant authentication (not to mention that likely JWT token checking can easily be the computationally most expensive part in an implementation, as you need to verify the signature)

**Comment by UlfBj** [comment]
Yes, the security comes with a cost. 
To come around that cost by relying on the notion of client sessions, seems not robust enough in my view. 
If a robust solution can be presented, it would be discussed in the WG. 

Using subscription, a client can receive multiple signals as a result of one request.

**Comment by SebastianSchildt** [comment]
I first would like to understand, what is the argument  behind the view that sessions are not robust enough? That is virtually done by "everyone", starting on lower layers like TLS up to (maybe more akin to VISS) logins to online shops, your bank, your cloud provider?

**Comment by UlfBj** [comment]
All the transport protocols supported in VISS must run over TLS, as I am sure you know. 
How long is a TLS session when it comes to HTTP? Does it extend over multiple HTTP requests?
Even if it does, is it good security practise to let the token based access control to "rely" on TLS?
I am asking because I am not sure about this.

**Comment by SebastianSchildt** [comment]
That was a misunderstanding: You are correct, TLS is only transport layer, as such it does not survive multiple HTTP requests (except HTTP pipelining features, to be more precise: It does not survive multiple TCP connections). I merely use it as an example of protocol, that for its purposes also establishes a "session" that remains valid as long as the connection is there.

On the layer (i.e. VISS) above you need session handling as well (as those "relations" can span multiple TLS establishments). For HTTPS (classic web stuff), this is done using session identifiers, potentially stored in cookies (so basically a much shorter random string, compared to a JWT token). For web/REST those time out, either after minutes or months, depending on the application.

For websocket it is even easier, as there you have the underlying TCP connection as long as websocket is "connected" (basically websocket emulates TCP semantics on a higher layer) . As such a "session" is implicitly established by opening a connection and teared down when the connection closes/breaks. So a server only needs to "remember" whether a client on a specific connection authenticated and how it is authorised without the need of a client to send any session identifier.

The only use case I came across where security credentials need to be self-contained in each "message" is in DTN (Delay-Tolerant Networking: Data-ferrying,Store-Carry-and forward, IPN ...), where your packet runtime is hours/days/months and no roundtrips/virtual connections are practically possible. But I think those are not VISS domains.

**Comment by UlfBj** [comment]
Yes, for HTTP session handling is not as straight forward as for WS. And I think the handling shall be uniform for the different transport protocols. But your mentioning of cookies could be a solution. E. g. the server could return a random nonce (the cookie) together with the signal value(s), and this could then be used instead of the token. Maybe we should bring this up for discussion in the WG?

**Comment by wonsuk73** [comment]
I think basically it's okay to me because the access control is optional for all of operations. But current structure of the spec described as below. It might make misunderstanding to readers. It looks only Read operation has authorized action even if other operations like Search read, History read and Service Discovery read also can be operated on the authorized channel. 

HTTPS:
5.1.2.1 Read
  5.1.2.1.1 Authorized Read
  5.1.2.1.2 Search Read
  5.1.2.1.3 History Read
  5.1.2.1.4 Service Discovery Read
5.1.2.2 Update
  5.1.2.2.1 Authorized Update

Websocket:
5.2.2.1 Read
  5.2.2.1.1 Authorized Read
  5.2.2.1.2 Search Read 
  5.2.2.1.3 History Read
  5.2.2.1.4 Service Discovery Read
5.2.2.2 Update
  5.2.2.2.1 Authorized Update
5.2.2.3 Subscribe
  5.2.2.3.1 Authorized Subscribe
  5.2.2.3.2 Curve Logging Subscribe
  5.2.2.4 Unsubscribe

If my comment is reasonable, how about revising this as below and adding the normal and authorized examples to each sub-section respectively?

HTTPS:
5.1.2.1 Read
  5.1.2.1.1 Basic Read
  5.1.2.1.2 Search Read
  5.1.2.1.3 History Read
  5.1.2.1.4 Service Discovery Read
5.1.2.2 Update
  5.1.2.2.1 Basic Update

Websocket:
5.2.2.1 Read
  5.2.2.1.1 Basic Read
  5.2.2.1.2 Search Read 
  5.2.2.1.3 History Read
  5.2.2.1.4 Service Discovery Read
5.2.2.2 Update
  5.2.2.2.1 Basic Update
5.2.2.3 Subscribe
  5.2.2.3.1 Basic Subscribe
  5.2.2.3.2 Curve Logging Subscribe
 5.2.2.4 Unsubscribe

**Comment by wonsuk73** [comment]
For long-lived HTTPS session using cookie I think we need to put this on the table for discussion in the group. Because I think the subscribe operation is a function typically used in the field.

**Comment by UlfBj** [comment]
>> If my comment is reasonable, how about revising this as below and adding the normal and authorized examples to each sub-section respectively?

@wonsuk73 I think your point is relevant, that it is not sufficiently clear how to use authorization in conjunction with the different request types. 
However, I think your proposed solution will be a bit "overkill". I believe a reader of this spec is able to generalise from the given example how to use it for different request types. But it should be made clear in the text of the existing example that authorization is supported in all request types. So adding this explanation I believe should be sufficient to solve the problem.

**Comment by tguild** [comment]
I encourage @SebastianSchildt and @isaacagudo to look at [call minutes](https://www.w3.org/2021/08/24-auto-minutes) for this topic. The group is supportive of convenience if not compromising security and then wonder what the mechanism should be to convey a shorter nonce (probably the 40 byte SHA256 signature for the token), is there something in Oauth2 we can use since we transmit the token via Authorization:Bearer?

**Comment by UlfBj** [comment]
To make the proposal to use only the signature part of the token in subsequent requests robust, the server must, after the first successful validation of the token, save the following data:
    - The token signature,
    - The token expiry time,
    - The signals that are claimed in the token, together with the access mode for each. 

The header currently defined to carry the token can also be used in the case only the signature part is sent as the server is able to identify whether the data represents a complete token, or only the signature part. A complete token must contain dots delimiting the different parts, while the data for the signature only does not contain any dots. 

A new error response is needed to inform a client that the server cannot grant authorisation without receiving a complete token. 
Error response: 401/token_incomplete/A complete token must be used. 

If we go with this proposal, I think we should use the full signature part of the token, in order to not reduce the security level that this signature provides. Its size is 43 bytes (32*1,33), while a complete VISSv2 JWT seems to be around 250-300 bytes (600-700% reduction).

This can be an optional feature, with the error response above being mandatory.

**Comment by SebastianSchildt** [comment]
If I fully got the proposal, I feel it needlessly weakens security: Normally for "session identifiers" it is quite elementary they are _random_ and cannot be easily guessed.

Of course the argument can be made, that stealing the token "signature" is as hard as stealing the complete token, but I still feel VISS should follow common security practices, and not use a fixed element, where it can/should be random.

https://xkcd.com/221/

Also, a "fixed" short form for a token is required to identify "sessions" in a more efficient way, if the server wants to do discern two nodes using the same token, it would need to include IP address (whereas still different nodes might connect from a single IP), and when moving to multipath capable protocols, it might get even harder. (A usecase where you want to discern individual connections might be rate limiting, if a client oversteps its bounds: Without an unambiguously identifiable endpoint, your only chance might be blocking the external IP of a proxy serving a xx k people company)

So in summary I feel the "short hand" for a token used by a specific client to identify a session should follow common practice and be a random "thing" selected by the server (i.e. an uuid would probably fit to the json VISS)

**Comment by SebastianSchildt** [comment]
Regarding OAuth, I have 0 experience with, but some googling, leads me to believe that the `nonce` might be what  you are searching: https://openid.net/specs/openid-connect-core-1_0.html 

> nonce
String value used to associate a Client session with an ID Token, and to mitigate replay attacks. The value is passed through unmodified from the Authentication Request to the ID Token. If present in the ID Token, Clients MUST verify that the nonce Claim Value is equal to the value of the nonce parameter sent in the Authentication Request. If present in the Authentication Request, Authorization Servers MUST include a nonce Claim in the ID Token with the Claim Value being the nonce value sent in the Authentication Request. Authorization Servers SHOULD perform no other processing on nonce values used. The nonce value is a case sensitive string.

**Comment by UlfBj** [comment]
>> If I fully got the proposal, I feel it needlessly weakens security: Normally for "session identifiers" it is quite elementary they are random and cannot be easily guessed.

How do mean the signature part of the token can be easily guessed?
If that is possible, then I believe you have broken the HMAC_SHA256 algo. 

>> Of course the argument can be made, that stealing the token "signature" is as hard as stealing the complete token, but I still feel VISS should follow common security practices, and not use a fixed element, where it can/should be random.

It ought to be as hard, which then would make the proposal as secure as keeping to send the entire token. If sending the entire token is not secure enough, then we have a problem. I cannot see what makes a random nonce generated by the server and returned to the client more secure than instead using the signature part of the token. Please enlighten me.

**Comment by SebastianSchildt** [comment]
Generally speaking it gives an attacker knowledge of parts of a packet (that is not the neccessarily the same as knowing the plaintext, knowing it is a a part of the token and will always be the same for a given client/token is already information)

Another reason for random elements choosen by one or both of the participants in a communication protocols is given in the the oauth snippet：Preventing replay attacks (in this case replaying messages from a previous session in a future one).

Also without further methods implemented by the server (some Ip adresses, ip,port tuples ASSUMING a client resues the same send port, or adding some tls, ws context for that cases), makes it hard to discern sessions (a session identifier is only good, if it is different for different sessions), would enable an attacker with the same token as me to impersonate me and for example  unsubscribing on my behalf. So if we have a bad apple, it can not „only“ attack the server, but also other peers. (This is different from e.g. Wifi: Even if I am evil and know your WiFi passphrase I can not listen to your computers Wifi connection (if I have not listened to/tampered with your handshake))

Please note, I do not have a security background, so mainly I try to lean on best practices invented by more knowledgeable people than me, so my strongest argument is, that I „get a bad feeling“ if some protocol/product deviates from standard security patterns.


In this specific case I would also argue, that from an implemtation point of view random identifiers are easier (or at least definitely not harder) to implement. And it would decouple the „session identifier“ from the format defined for Jwt tokens, i.e. if we later define them or allow them  to use another MAC, that again may grow uncomfortably large

**Comment by isaacagudo** [comment]
I am not sure I completely understand the issue but I'll try to describe some scenarios and how I think the protocol should behave in general.

1. Client establish a TLS connection and makes all requests under the same TLS connection either using wss:// or https://, o even mqtts...

In this case, client authentication only needs to happen onces, right after the TLS connections is stablished. Regarding authorization, as we don't support any kind of dynamic revocation of access rights, the authorization information only need to be resend once they expire. The server can cache this information and link it to the TLS session while it is valid. 

2. Client need to reconnect using TLS often.

In this case, each time the client needs to reconnect using TLS is has to authenticate again after the new TLS connection is stablished. We assume the TLS connection is not resumed and we can't take advantage of any quick reconnection as it will restore the previous TLS session making this equivalent to the first scenario. Regarding authorization, once the client is authenticated we can use the previously cached information and link it to the new session.

3. Client has no access using TLS

In this case, we could still make it work but the client need to authenticate in each request using some challenge response mechanisms that prevents an eavesdropper from reusing previous authentication exchanges to impersonate the client. Regarding authorization, we need to attach this information to the authentication request, in a way that it is of no use for eavesdroppers, although we only need to send it after they expire in the server, i.e. not in each authentication request.

This three scenarios assume the server has a means to store connected clients information and their corresponding access rights, but in general we could have also here different scenarios:

1. Server has no caching capabilities at all.

In this case the 3 scenarios above collapse into one and there is no workaround, the client need to authenticate in each request and either it also pass the authorization information in every request or the server retrieves it from the Access Grant Server ...

2. Server can only cache limited access rights per client.

In this case, if the client need to elevate or switch roles it need to resend authorization information.

**Implementing unlimited caching would be risky and would open the door to DoS attacks.**

In any case, the key point for this discussion is how the client is authenticated and how is authorization information sent to the server within the specification. In general we have left the discussion about how to authenticate clients outside of the specification, but this decision depends on wether we use short term or long term AGTs.

1. When using short term AGT, client authentication is not really covered in the specification but we are taking the same approach as OAuth 2.0, [The OAuth 2.0 Authorization Framework: Bearer Token Usage](https://datatracker.ietf.org/doc/html/rfc6750), i.e. give access to the bearer of this token. The AT both authenticates the client and carries access right with it, we can't separate them as it is currently defined. In this case every time the user needs to re-authenticate or refresh the authorization information it need to send a whole AT.

2. When using long term AGT, the client has a secret key that can be used to authenticate itself without having to resend any extra information. It can do that because the matching public key is included in the AGT. In case the AGT is sent to the server, it will have a mechanism to authenticate the client using public key cryptography. Then, the client could just sign a nonce/timestamp in a challenge/response protocol and be authenticated by the server without resending the AT. Regarding authorization, every time its needs to be renewed a whole AT will need to we send.

I hope this discussion helps us understanding better the security requirements and setups under different circumstances. We could elaborate more on that in the next meeting.

Regarding the actual questions in the issue, I think it seems reasonable to have a simpler token just for authentication but this will increase the complexity of the specification because we will be no longer able to leave client authentication outside the specification. Still we need to decide who issues this simple token, will it be the VISS (similar to a cookie) or the AT server (similar to OAuth) ...

**Comment by wonsuk73** [comment]
>However, I think your proposed solution will be a bit "overkill". I believe a reader of this spec is able to generalise from the given example how to use it for different request types. But it should be made clear in the text of the existing example that authorization is supported in all request types. So adding this explanation I believe should be sufficient to solve the problem.

@UlfBj I understand your point. But to make clear this part in the spec I think we might need to revise the section of 5.1.2.1 Read. Below is current structure for HTTPS transport. In my view separated sub-section for authorized read is confuse. I have two proposals as below. I don't have a strong opinion on these but I have to choose I might do proposal 1.

Proposal 1) 
Texts in Authorized Read are merged after example in of 5.1.2.1 Read and remove 5.1.2.1.1 Authorized Read. Because other operation sections as search, history and service discovery might need to cover authorized read with additional texts even not an example. 

Proposal 2) 
Another way is just left common texts for 4 different read operation in 5.1.2.1 Read. And move an example of Read operation in to 5.1.2.1.1 and change the title of this section as Basic Read.

HTTPS:
5.1.2.1 Read
   5.1.2.1.1 Authorized Read
   5.1.2.1.2 Search Read
   5.1.2.1.3 History Read
   5.1.2.1.4 Service Discovery Read
5.1.2.2 Update
   5.1.2.2.1 Authorized Update

What do you think? Dose it make sense?

**Comment by UlfBj** [comment]
@wonsuk73 : Please have a look at the PR#406, which aims at resolving this.

**Comment by wonsuk73** [comment]
@UlfBj Thanks for the text updates for this issue. I think it might need to update sentenses for Authorized Read and Authorized Update on Websocket transport. I made a PR for for this as PR #411. Please review this.

**Comment by UlfBj** [comment]
Fixed in PR411
